sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -40,6 +40,13 @@ abstract class LogicalPlan
     super.verboseString(maxFields) + statsCache.map(", " + _.toString).getOrElse("")
   }
 
+  override protected def doCanonicalize(): LogicalPlan = {
+    if (!resolved) {
```

Comment:
Is it an assert rather than throwing an exception?

Reply:
Considered use assert, but it can be happen in a special way. e.g.
```
spark.sql("select id, count(*) from t1 group by id limit 1").queryExecution.logical.canonicalized
```

So throwing maybe better.

Reply:
You think users use canonicalization? I think that is for internal use only though.